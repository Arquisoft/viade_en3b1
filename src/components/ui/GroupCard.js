import React from 'react';
import { Card, CardContent, IconButton, CardMedia, Typography, makeStyles, CardActions, Collapse, Link } from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import clsx from 'clsx';
import PeopleRoundedIcon from '@material-ui/icons/PeopleRounded';

function GroupCard(props) {
    const classes = useStyles();
    const { group } = props;
    const members = group.getMembers();

    const [expanded, setExpanded] = React.useState(false);

    const handleExpandClick = () => {
        setExpanded(!expanded);
    };

    function getName(webid) {
        var username = webid;
        username = username.replace('https://', '');
        username = username.replace('.solid.community/profile/card#me', '');
        username = username.replace('.inrupt.net/profile/card#me', '');
        return username;
    }


    return (
        <div>
            <Card elevation={4} className={classes.root}>
                <div className={classes.details}>
                    <CardContent className={classes.content}>
                        <Typography className={classes.title} component="h5" variant="h5">
                            {group.getName()}
                        </Typography>
                    </CardContent>

                    <CardActions disableSpacing>
                        <IconButton
                            data-testid="btn-expand-test"
                            className={clsx(classes.expand, {
                                [classes.expandOpen]: expanded,
                            })}
                            onClick={handleExpandClick}
                            aria-expanded={expanded}
                            aria-label="show more"
                        >
                            <ExpandMoreIcon />
                        </IconButton>
                    </CardActions>
                    <Collapse in={expanded} timeout="auto" unmountOnExit>
                        <CardContent >
                            {members.map((each, index) => (
                                <div key={index}>
                                    <Link href={each} target="_blank">{getName(each)}</Link>
                                </div>
                            ))}
                        </CardContent>
                    </Collapse>
                </div>
                <CardMedia
                    className={classes.cover}
                >
                    <PeopleRoundedIcon style={{fontSize: '5rem', fill: '#7c4dff'}} />
                </CardMedia>
            </Card>
        </div>
    );
}

const useStyles = makeStyles((theme) => ({
    root: {
        display: 'flex',
        marginTop: theme.spacing(5)
    },
    details: {
        display: 'flex',
        flexDirection: 'column',
        width: theme.spacing(50)
    },
    content: {
        flex: '1 0 auto',
    },
    title: {
        martinTop: theme.spacing(5)
    },
    cover: {
        width: theme.spacing(30),
        marginLeft: theme.spacing(12),
        marginTop: theme.spacing(3)
    },
    expand: {
        transform: 'rotate(0deg)',
        transition: theme.transitions.create('transform', {
            duration: theme.transitions.duration.shortest,
        }),
    },
    expandOpen: {
        transform: 'rotate(180deg)',
    },
}));

export default GroupCard;