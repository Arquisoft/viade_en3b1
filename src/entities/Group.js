import { v4 as uuid } from "uuid";

class Group {

    constructor(name, members = [], id = null) {
        this.name = name;
        this.members = members;

        if (id === null) {
            this.id = uuid().toString();
        } else {
            this.id = id;
        }
    }

    getName() {
        return this.name;
    }

    getMembers() {
        return this.members;
    }

    getId() {
        return this.id;
    }

    toJsonLD() {
        let membersJSON = [];
        this.members.forEach((url) => {
            membersJSON.push(({
                "url": url
            }));
        });
        return JSON.stringify(
            {
                "@context": {
                    "@version": 1.1,
                    "users": {
                        "@container": "@list",
                        "@id": "schema:Person"
                    },
                    "name": {
                        "@id": "schema:name",
                        "@type": "xs:string"
                    },
                    "url": {
                        "@id": "schema:url",
                        "@type": "xs:string"
                    },
                    "schema": "http://schema.org/",
                    "xsd": "http://www.w3.org/2001/XMLSchema#"
                },
                "id": this.id,
                "name": this.name,
                "users": membersJSON
            }
        );
    }

}

export default Group;