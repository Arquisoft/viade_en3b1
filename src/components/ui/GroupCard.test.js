import GroupCard from './GroupCard';
import Group from '../../entities/Group';
import { render, fireEvent, getByTestId } from '@testing-library/react';

var group1 = new Group("Grupo1", ["url1", "url2"]);

test('Define GroupCard', () => {
  expect(GroupCard).toBeDefined();
});

test('Renders correctly', () => {
  const wrapper = render(<GroupCard group={group1} />);
  expect(wrapper).toMatchSnapshot();
});

test('Click on expand', () => {
    const { container } = render(<GroupCard group={group1}/>);

    const button = getByTestId(container, 'btn-expand-test');
    fireEvent.click(button);
});