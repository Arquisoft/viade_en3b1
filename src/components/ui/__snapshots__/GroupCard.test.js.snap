// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation4 MuiPaper-rounded"
        >
          <div
            class="makeStyles-details-2"
          >
            <div
              class="MuiCardContent-root makeStyles-content-3"
            >
              <h5
                class="MuiTypography-root makeStyles-title-4 MuiTypography-h5"
              >
                Grupo1
              </h5>
            </div>
            <div
              class="MuiCardActions-root"
            >
              <button
                aria-expanded="false"
                aria-label="show more"
                class="MuiButtonBase-root MuiIconButton-root makeStyles-expand-6"
                data-testid="btn-expand-test"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
          <div
            class="MuiCardMedia-root makeStyles-cover-5"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              style="font-size: 5rem; fill: #7c4dff;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16 11c1.66 0 2.99-1.34 2.99-3S17.66 5 16 5s-3 1.34-3 3 1.34 3 3 3zm-8 0c1.66 0 2.99-1.34 2.99-3S9.66 5 8 5 5 6.34 5 8s1.34 3 3 3zm0 2c-2.33 0-7 1.17-7 3.5V18c0 .55.45 1 1 1h12c.55 0 1-.45 1-1v-1.5c0-2.33-4.67-3.5-7-3.5zm8 0c-.29 0-.62.02-.97.05.02.01.03.03.04.04 1.14.83 1.93 1.94 1.93 3.41V18c0 .35-.07.69-.18 1H22c.55 0 1-.45 1-1v-1.5c0-2.33-4.67-3.5-7-3.5z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation4 MuiPaper-rounded"
      >
        <div
          class="makeStyles-details-2"
        >
          <div
            class="MuiCardContent-root makeStyles-content-3"
          >
            <h5
              class="MuiTypography-root makeStyles-title-4 MuiTypography-h5"
            >
              Grupo1
            </h5>
          </div>
          <div
            class="MuiCardActions-root"
          >
            <button
              aria-expanded="false"
              aria-label="show more"
              class="MuiButtonBase-root MuiIconButton-root makeStyles-expand-6"
              data-testid="btn-expand-test"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <div
          class="MuiCardMedia-root makeStyles-cover-5"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            style="font-size: 5rem; fill: #7c4dff;"
            viewBox="0 0 24 24"
          >
            <path
              d="M16 11c1.66 0 2.99-1.34 2.99-3S17.66 5 16 5s-3 1.34-3 3 1.34 3 3 3zm-8 0c1.66 0 2.99-1.34 2.99-3S9.66 5 8 5 5 6.34 5 8s1.34 3 3 3zm0 2c-2.33 0-7 1.17-7 3.5V18c0 .55.45 1 1 1h12c.55 0 1-.45 1-1v-1.5c0-2.33-4.67-3.5-7-3.5zm8 0c-.29 0-.62.02-.97.05.02.01.03.03.04.04 1.14.83 1.93 1.94 1.93 3.41V18c0 .35-.07.69-.18 1H22c.55 0 1-.45 1-1v-1.5c0-2.33-4.67-3.5-7-3.5z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
